### PR TITLE
feat(storage-plugin): add before and after serialize hooks

### PR DESCRIPTION
**Pull request by davidmorissette**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

There is no per-state serialization.

Issue Number: N/A

## What is the new behavior?

Per-state serialization offers full control on how states are serialize/deserialize.

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

## Other information

This PR is different from the first one as it leverages the serialize/deserialize functions from the plugin options.

### CODE REVIEW

File: packages/storage-plugin/src/internals.ts [hunk 1]
```diff
@@ -53,8 +53,8 @@ export function storageOptionsFactory(
   return {
     key: [DEFAULT_STATE_KEY],
     storage: StorageOption.LocalStorage,
-    serialize: JSON.stringify,
-    deserialize: JSON.parse,
+    serialize: obj => JSON.stringify(obj),
```
Comment (splincode) [inline packages/storage-plugin/src/internals.ts]: it's breaking change?

**Comment by davidmorissette** [comment]
@splincode These are the defaults for serialize/deserialize. I had to update the way JSON.* functions are called since the new interface takes an optional `key` parameter. My goal was to NOT introduce a breaking change since the `key` parameter is optional, so existing code should still work as is.

**Comment by markwhitfeld** [comment]
I think that the change to add the `key` parameter is great. But this may break existing code (for the same reason that you were forced to change the JSON. * function calls). I think that we should define a method alternative on the interface to avoid this. Ill add a suggestion.

@arturovt and @splincode my first thought is that I don't want to add an additional api for this serialisation strategy in the course of the plugin. It could be a labs extension. What do you think?

File: packages/storage-plugin/src/symbols.ts [hunk 2]
```diff
@@ -46,14 +46,14 @@ export interface NgxsStoragePluginOptions {
   }[];
 
   /**
-   * Serailizer for the object before its pushed into the engine.
+   * Serializer for the object before its pushed into the engine.
    */
-  serialize?(obj: any): string;
+  serialize?(obj: any, key?: string): string;
```
Comment (markwhitfeld) [inline packages/storage-plugin/src/symbols.ts]:
```suggestion
  serialize?(obj: any): string;
  serialize?(obj: any, key: string): string;
```

File: packages/storage-plugin/src/symbols.ts [hunk 3]
```diff
@@ -46,14 +46,14 @@ export interface NgxsStoragePluginOptions {
...
 
   /**
    * Deserializer for the object before its pulled out of the engine.
    */
-  deserialize?(obj: any): any;
+  deserialize?(obj: any, key?: string): any;
```
Comment (markwhitfeld) [inline packages/storage-plugin/src/symbols.ts]:
```suggestion
  deserialize?(obj: any): any;
  deserialize?(obj: any, key: string): any;
```

File: packages/storage-plugin/src/internals.ts [hunk 4]
```diff
@@ -53,8 +53,8 @@ export function storageOptionsFactory(
...
+    deserialize: obj => JSON.parse(obj),
```
Comment (markwhitfeld) [inline packages/storage-plugin/src/internals.ts]: Once you have done the recommendation below please see if you can revert these changes and it is happy with it

**Comment by splincode** [comment]
> my first thought is that I don't want to add an additional api for this serialisation strategy in the course of the plugin. It could be a labs extension. What do you think?

Not sure

**Comment by davidmorissette** [comment]
@splincode @markwhitfeld I added to new test "Should not introduce any breaking changes." to show you that my changes do not introduce any breaking changes. In my example, the `key` parameter is not used.

**Comment by splincode** [comment]
@davidmorissette any progress?

**Comment by davidmorissette** [comment]
@splincode I was waiting on you guys to see if the test I added "Should not introduce any breaking changes." demonstrated that PR was not introducing any breaking changes. What's is your feedback?

**Comment by splincode** [comment]
@markwhitfeld ping

**Comment by davidmorissette** [comment]
@markwhitfeld @splincode What is the status on this PR right now?

**Comment by splincode** [comment]
@davidmorissette have you fixed in the code by comments?

**Comment by davidmorissette** [comment]
@splincode I added a test "Should not introduce any breaking changes." before changing anything. Can you have a look so we can discuss about it?

File: packages/storage-plugin/src/symbols.ts (hunk 3 above)
Comment (splincode) [reply packages/storage-plugin/src/symbols.ts]: @davidmorissette please revert it

File: packages/storage-plugin/src/symbols.ts (hunk 2 above)
Comment (splincode) [reply packages/storage-plugin/src/symbols.ts]: @davidmorissette please add

```ts
serialize?(obj: any): any;
serialize?(obj: any, key: string): any;
```

File: packages/storage-plugin/src/internals.ts (hunk 4 above)
Comment (splincode) [reply packages/storage-plugin/src/internals.ts]: @davidmorissette see

**Comment by splincode** [comment]
`Should not introduce any breaking changes.` - good test

**Comment by markwhitfeld** [comment]
Sorry for the silence on my end. I have been crazy busy with a number of other things.
I think that the change to add the key is great (glad we have a test to prove that it is not a breaking change).

Unfortunately, the serialization strategy part of this PR is what I would not want to add to the plugin. It increases the plugin size by 1.5kb for ES6 and 3kb for ES5.
It also adds an additional API that we would have to maintain backward compatibility for.
I was asking the other core team members what they thought of it in my last https://github.com/ngxs/store/pull/1513#issuecomment-573332888:
> @arturovt and @splincode my first thought is that I don't want to add an additional api for this serialization strategy in the course of the plugin. It could be a labs extension. What do you think?

@davidmorissette if you remove the serialization strategy stuff from this PR and just leave the changes for the key then I would be happier to accept.
The strategy stuff could be a great labs plugin. 
We can even reference it from the storage plugin docs.

File: packages/storage-plugin/tests/storage.plugin.spec.ts [hunk 5]
```diff
@@ -462,4 +478,99 @@ describe('NgxsStoragePlugin', () => {
...
+
+      const deserialize = (obj: any): any => JSON.parse(obj);
+
+      // Act
+      TestBed.configureTestingModule({
+        imports: [
+          NgxsModule.forRoot([CounterState]),
+          NgxsStoragePluginModule.forRoot({
+            key: 'counter',
+            deserialize: deserialize
```
Comment (markwhitfeld) [inline packages/storage-plugin/tests/storage.plugin.spec.ts]: Please change this line as follows:
```suggestion
            deserialize: JSON.parse
```
If this change works then I would be confident that this is not a breaking change.

File: packages/storage-plugin/src/symbols.ts (hunk 3 above)
Comment (markwhitfeld) [inline packages/storage-plugin/src/symbols.ts]: In order to maintain backward compatibility we would need to change to this:
```suggestion
  deserialize?(obj: any): any;
  deserialize?(obj: any, key?: string): any;
```
I did a test.
This does not work:
image
It gives this error:
image

But, this works:
image

**Comment by splincode** [comment]
Please rebase from master

File: packages/storage-plugin/src/symbols.ts (hunk 3 above)
Comment (davidmorissette) [reply packages/storage-plugin/src/symbols.ts]: @markwhitfeld Not sure if you can manage to get it work that way. I tried your example and the compiler was unable to see the second method. Maybe I am missing something with my understanding of overloads in Typescript.

https://www.typescriptlang.org/docs/handbook/functions.html#overloads

I used a slightly different approach. I kept the existing serialize/deserialize methods untouched and I added optional beforeSerialize/afterDeserialize methods. With that approach, not breaking existing code is guaranteed and it offers a better separation of concerns. The serialize/deserialize are about how to serialize and deserialize the objects and the beforeSerialize and afterSerialize are about how to alter these objects before of after the serialization itself.

I am waiting on your feedback.

Regarding whether or not you want to include the 'SerializationStrategy' class in the codebase, I will respect your decision. Having the opportunity to extends the functionality thanks to the new beforeSerialize/afterDeserialize is good enough for me. Please keep me updated so I can remove it and adjust the tests.

**Comment by splincode** [comment]
@davidmorissette Could you correctly cringe with the master? To show only new changes

**Comment by davidmorissette** [comment]
@splincode Sorry I don't really know how to do it since I already rebased and pushed.

**Comment by markwhitfeld** [comment]
Hi @davidmorissette 
Your rebase really broke the history for this PR. I think it is because it included a merge within its history. It makes this PR impossible to review.
I have made a repaired branch based on your changes (I did some cherry picking).

Could you fetch my repo and set your branch to this commit and then push it to your branch that this PR is based on:
https://github.com/markwhitfeld/store/tree/feat/serialization-key
I think that will fix it.

I like the idea of your before and after methods.
Also, I would not want to include your serialization strategy in the standard library, but this change should enable you to have that in your project (or to create a labs project for it).

**Comment by davidmorissette** [comment]
@markwhitfeld Since the amount of code was small and I had to make some changes in order to remove the SerializationStrategy class, I opted to reset my branch to the point before all my commits and to start over. It should be easier to read now.

File: package.json [hunk 6]
```diff
@@ -110,7 +110,7 @@
   "private": true,
   "devDependencies": {
     "@angular-builders/jest": "^8.3.2",
-    "@angular-devkit/build-angular": "~0.803.24",
+    "@angular-devkit/build-angular": "~0.900.4",
```
Comment (splincode) [inline package.json]: please revert it

File: package.json [hunk 7]
```diff
@@ -125,7 +125,7 @@
     "@angular/router": "8.2.14",
     "@commitlint/cli": "^8.3.5",
     "@commitlint/config-conventional": "^8.3.4",
-    "@cypress/webpack-preprocessor": "^4.1.1",
+    "@cypress/webpack-preprocessor": "^4.1.3",
```
Comment (splincode) [inline package.json]: please revert it

File: package.json [hunk 8]
```diff
@@ -155,7 +155,7 @@
     "ts-loader": "^6.2.1",
     "ts-node": "^8.6.2",
     "tsickle": "^0.38.0",
-    "tslib": "^1.10.0",
+    "tslib": "^1.11.1",
```
Comment (splincode) [inline package.json]: please revert it

File: .gitignore [hunk 9]
```diff
@@ -26,6 +26,10 @@
 dist-integration
 /dist
 
+# IDE - VS
```
Comment (splincode) [inline .gitignore]: please revert it (add to another pr)

**Comment by splincode** [comment]
@markwhitfeld I think good

**Comment by davidmorissette** [comment]
@markwhitfeld Can you please provide some guidance about how to create a lab project so I can share the rest of my work? Thank you!

**Comment by markwhitfeld** [comment]
@splincode Could you assist David with this?

**Comment by splincode** [comment]
@davidmorissette done
https://github.com/ngxs-labs/storage-plugin-extension

you have full permission for current repo

**Comment by vinagreti** [comment]
Guys? Any news? The `storage-plugin-extension`  repo is empty :(
I would be very happy if it gets to the production version :D

**Comment by davidmorissette** [comment]
@splincode Could you create @ngxs-labs/storage-plugin-extension (npm) and add me?

**Comment by splincode** [comment]
@davidmorissette 

image

I did it on March 14

**Comment by davidmorissette** [comment]
@splincode I was referring to npm, not GitHub. Is there anything you need to setup prior to publish to the public scope "@ngxs-labs" ?

**Comment by splincode** [comment]
Before giving you rights to NPM, you need to push your changes to the repository so that we can evaluate it

**Comment by davidmorissette** [comment]
@vinagreti 

> Guys? Any news? The storage-plugin-extension  repo is empty :(
> I would be very happy if it gets to the production version :D

I am waiting on the next release of the storage-plugin so I can build the extension plugin on top of it.

@markwhitfeld Do you have a general idea on when this PR will be released?

**Comment by markwhitfeld** [comment]
Hi @davidmorissette. I would like to do a release soon. Do you think that you could add a section on this feature into the announcement article for me?
https://github.com/ngxs/store/blob/master/publication/announcing_ngxs_next/article.md

**Comment by davidmorissette** [comment]
@markwhitfeld Please have a look to my PR #1587

**Comment by davidmorissette** [comment]
@markwhitfeld @splincode It has been a while! I finally pushed the Storage Plugin Extension. The first feature on this plugin is the InterceptorStrategy class which offers a convenient way to configure before and after serializations on the Storage Plugin. Please have a look on the solution and go ahead on performing the configuration/installation required within the labs. Thank you!
